src/main/java/org/tron/common/runtime/vm/program/VMTransaction.java

```diff
@@ -198,194 +207,214 @@ public synchronized void protoParse() {
-     */
-    public void sign(byte[] privKeyBytes) throws MissingPrivateKeyException {
-        sign(ECKey.fromPrivate(privKeyBytes));
+      //use protoEncoded directly and it should never be null
+      this.hash = Hash.sha3(protoEncoded);
+      this.parsed = true;
+      TransactionCapsule transaction = new TransactionCapsule(protoEncoded);
+    } catch (Exception e) {
+      throw new RuntimeException("Error on parsing proto", e);
+    }
+  }
+
+  private void validate() {
+    if (receiveAddress != null && receiveAddress.length != 0
+        && receiveAddress.length != ADDRESS_LENGTH) {
+      throw new RuntimeException("Receive address is not valid");
     }
 
-    public void sign(ECKey key) throws MissingPrivateKeyException {
-        this.signature = key.sign(this.getRawHash());
-        this.protoEncoded = null;
-    }
-
-    @Override
-    public String toString() {
-        return toString(Integer.MAX_VALUE);
-    }
-
-    public String toString(int maxDataSize) {
-        protoParse();
-        String dataS;
-        if (data == null) {
-            dataS = "";
-        } else if (data.length < maxDataSize) {
-            dataS = ByteUtil.toHexString(data);
-        } else {
-            dataS = ByteUtil.toHexString(Arrays.copyOfRange(data, 0, maxDataSize)) +
-                    "... (" + data.length + " bytes)";
-        }
-        return "TransactionData [" + "hash=" + ByteUtil.toHexString(hash) +
-                ", receiveAddress=" + ByteUtil.toHexString(receiveAddress) +
-                ", sendAddress=" + ByteUtil.toHexString(getSender())  +
-                ", value=" + value +
-                ", data=" + dataS +
-                ", signatureV=" + (signature == null ? "" : signature.v) +
-                ", signatureR=" + (signature == null ? "" : ByteUtil.toHexString(BigIntegers.asUnsignedByteArray(signature.r))) +
-                ", signatureS=" + (signature == null ? "" : ByteUtil.toHexString(BigIntegers.asUnsignedByteArray(signature.s))) +
-                "]";
-    }
-
-    /**
-     * For signatures you have to keep also
-     * RLP of the transaction without any signature data
-     */
-    public byte[] getEncodedRaw() {
-        protoParse();
-        if (protoRaw != null) return protoRaw;
-
-        return this.protoEncoded.clone();
+    if (getSignature() != null) {
+      if (BigIntegers.asUnsignedByteArray(signature.r).length > HASH_LENGTH) {
+        throw new RuntimeException("Signature R is not valid");
+      }
+      if (BigIntegers.asUnsignedByteArray(signature.s).length > HASH_LENGTH) {
+        throw new RuntimeException("Signature S is not valid");
+      }
+      if (getSender() != null && getSender().length != ADDRESS_LENGTH) {
+        throw new RuntimeException("Sender is not valid");
+      }
+    }
+  }
+
+  public boolean isParsed() {
+    return parsed;
+  }
+
+  public byte[] getHash() {
+    if (!isEmpty(hash)) {
+      return hash;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Return a copy of "hash". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1188&issues=AWTf_bGjrFWWQMKkhNw-&open=AWTf_bGjrFWWQMKkhNw-)

[ISSUE_KEY]: AWTf_bGjrFWWQMKkhNw-